File: lib/src/prom/queries.dart
import '../config.dart';

enum InstantQuery {
  up,
  cpuBusy,
  cores,
  memTotal,
  memAvailable,
  swapTotal,
  swapFree,
  load1,
  load5,
  load15,
  bootTime,
  fsSize,
  fsAvail,
  netRx,
  netTx,
  allTemps,
  cpuIoWait,
  gpuTemp,
  gpuUtil,
  gpuFbUsed,
  gpuFbFree,
  gpuPower,
  gpuSmClock,
  gpuMemClock,
  gpuMemTemp,
  gpuAgeDeep,
  gpuAgeFresh,
}

/// Every PromQL expression the app issues, in one place.
///
/// The target Prometheus scrapes `job="node"` (node_exporter on the Proxmox
/// host plus each guest) and `job="dcgm"` (dcgm-exporter next to the GPU).
abstract final class Q {
  // ---- per-instance node metrics -------------------------------------------

  static const cpuBusy =
      '100 - (avg by (instance) (rate(node_cpu_seconds_total{mode="idle"}[2m])) * 100)';
  static const cpuIoWait =
      'avg by (instance) (rate(node_cpu_seconds_total{mode="iowait"}[2m])) * 100';
  static const cores =
      'count by (instance) (node_cpu_seconds_total{mode="idle"})';

  static const memTotal = 'node_memory_MemTotal_bytes';
  static const memAvailable = 'node_memory_MemAvailable_bytes';
  static const swapTotal = 'node_memory_SwapTotal_bytes';
  static const swapFree = 'node_memory_SwapFree_bytes';

  static const load1 = 'node_load1';
  static const load5 = 'node_load5';
  static const load15 = 'node_load15';
  static const bootTime = 'node_boot_time_seconds';

  static const fsSize = 'node_filesystem_size_bytes{mountpoint="/"}';
  static const fsAvail = 'node_filesystem_avail_bytes{mountpoint="/"}';

  static final netRx =
      'sum by (instance) (rate(node_network_receive_bytes_total{device!~"${AppConfig.netDeviceExclude}"}[2m]))';
  static final netTx =
      'sum by (instance) (rate(node_network_transmit_bytes_total{device!~"${AppConfig.netDeviceExclude}"}[2m]))';

  static const up = 'up';

  /// Labelled hwmon temperatures for readable names such as `Tctl`/`Tccd1`.
  static const cpuTemp =
      'node_hwmon_temp_celsius * on(instance,chip,sensor) group_left(label) node_hwmon_sensor_label';

  /// All hwmon temperatures, with readable labels where node_exporter has one
  /// and the raw sensor id retained otherwise. `or on(...)` acts as a left join:
  /// the labelled series wins, while unmatched raw channels remain visible.
  static const allTemps =
      '($cpuTemp) or on(instance,chip,sensor) node_hwmon_temp_celsius';

  // ---- GPU (dcgm-exporter) -------------------------------------------------
  //
  // Live queries explicitly reject samples older than gpuStaleAfter. Prometheus
  // instant selectors otherwise keep returning the last sample from their
  // lookback window, which can make a field look live after it disappeared.
  // Seven-day fallback queries are separate: MetricsStore only executes them
  // for down exporters and caches the result for gpuFallbackRefresh.

  static const _gpuWindow = '7d';

  static String _freshGpu(String metric) {
    final fresh = AppConfig.gpuStaleAfter.inSeconds;
    return '$metric and on(instance,gpu,UUID) '
        '(time() - timestamp($metric) < $fresh)';
  }

  static String _lastGpu(String metric) =>
      'last_over_time($metric[$_gpuWindow])';

  static final gpuTemp = _freshGpu('DCGM_FI_DEV_GPU_TEMP');
  static final gpuMemTemp = _freshGpu('DCGM_FI_DEV_MEMORY_TEMP');
  static final gpuUtil = _freshGpu('DCGM_FI_DEV_GPU_UTIL');
  static final gpuFbUsed = _freshGpu('DCGM_FI_DEV_FB_USED');
  static final gpuFbFree = _freshGpu('DCGM_FI_DEV_FB_FREE');
  static final gpuPower = _freshGpu('DCGM_FI_DEV_POWER_USAGE');
  static final gpuSmClock = _freshGpu('DCGM_FI_DEV_SM_CLOCK');
  static final gpuMemClock = _freshGpu('DCGM_FI_DEV_MEM_CLOCK');

  static final gpuTempLast = _lastGpu('DCGM_FI_DEV_GPU_TEMP');
  static final gpuMemTempLast = _lastGpu('DCGM_FI_DEV_MEMORY_TEMP');
  static final gpuUtilLast = _lastGpu('DCGM_FI_DEV_GPU_UTIL');
  static final gpuFbUsedLast = _lastGpu('DCGM_FI_DEV_FB_USED');
  static final gpuFbFreeLast = _lastGpu('DCGM_FI_DEV_FB_FREE');
  static final gpuPowerLast = _lastGpu('DCGM_FI_DEV_POWER_USAGE');
  static final gpuSmClockLast = _lastGpu('DCGM_FI_DEV_SM_CLOCK');
  static final gpuMemClockLast = _lastGpu('DCGM_FI_DEV_MEM_CLOCK');

  /// Seconds since the GPU exporter last produced a temperature sample while
  /// that series remains within Prometheus' normal lookback window.
  static const gpuAgeFresh = 'time() - timestamp(DCGM_FI_DEV_GPU_TEMP)';

  /// Long-term age fallback for an exporter that has been down for hours/days.
  static const gpuAgeDeep =
      'time() - max_over_time(timestamp(DCGM_FI_DEV_GPU_TEMP)[$_gpuWindow:5m])';

  // ---- range queries for the GRAPHS screen ---------------------------------

  static const rangeCpu = cpuBusy;
  static const rangeMemPct =
      '(1 - (node_memory_MemAvailable_bytes / node_memory_MemTotal_bytes)) * 100';
  static const rangeTempCpu = cpuTemp;
  static const rangeTempGpu = 'DCGM_FI_DEV_GPU_TEMP';
  static const rangeGpuUtil = 'DCGM_FI_DEV_GPU_UTIL';
  static final rangeNetRx = netRx;

  /// Throughput measured end to end, one series per `path` (direct vs socks),
  /// rather than the interface counters — those say how busy the NIC was, not
  /// what the link is actually worth right now.
  ///
  /// The exporter runs on its own schedule, tens of minutes apart, so the raw
  /// metric is a scatter of isolated points over any window short enough to be
  /// interesting. `last_over_time` holds each result until the next run, which
  /// is what the number means anyway: the last measurement still stands.
  static const rangeSpeedtestDown =
      'last_over_time(speedtest_download_bits_per_second[1h])';

  /// Same treatment for the other direction. The exporter measures both in one
  /// run, so an upload point exists for every download point and the two share
  /// an axis honestly.
  static const rangeSpeedtestUp =
      'last_over_time(speedtest_upload_bits_per_second[1h])';

  static const rangeSpeedtestLatency =
      'last_over_time(speedtest_latency_seconds[1h])';

  /// All instant expressions used by a complete MetricsStore poll. The live
  /// smoke test consumes this same list so it cannot silently drift behind the
  /// production query batch.
  static final Map<InstantQuery, String> instantPollQueries = {
    InstantQuery.up: up,
    InstantQuery.cpuBusy: cpuBusy,
    InstantQuery.cores: cores,
    InstantQuery.memTotal: memTotal,
    InstantQuery.memAvailable: memAvailable,
    InstantQuery.swapTotal: swapTotal,
    InstantQuery.swapFree: swapFree,
    InstantQuery.load1: load1,
    InstantQuery.load5: load5,
    InstantQuery.load15: load15,
    InstantQuery.bootTime: bootTime,
    InstantQuery.fsSize: fsSize,
    InstantQuery.fsAvail: fsAvail,
    InstantQuery.netRx: netRx,
    InstantQuery.netTx: netTx,
    InstantQuery.allTemps: allTemps,
    InstantQuery.cpuIoWait: cpuIoWait,
    InstantQuery.gpuTemp: gpuTemp,
    InstantQuery.gpuUtil: gpuUtil,
    InstantQuery.gpuFbUsed: gpuFbUsed,
    InstantQuery.gpuFbFree: gpuFbFree,
    InstantQuery.gpuPower: gpuPower,
    InstantQuery.gpuSmClock: gpuSmClock,
    InstantQuery.gpuMemClock: gpuMemClock,
    InstantQuery.gpuMemTemp: gpuMemTemp,
    InstantQuery.gpuAgeFresh: gpuAgeFresh,
  };

  /// Expensive historical queries, only used for dcgm instances whose `up` is
  /// currently zero. Kept separate from [instantPollQueries] so the store can
  /// cache them and avoid seven-day scans on every poll.
  static final Map<InstantQuery, String> gpuFallbackQueries = {
    InstantQuery.gpuTemp: gpuTempLast,
    InstantQuery.gpuUtil: gpuUtilLast,
    InstantQuery.gpuFbUsed: gpuFbUsedLast,
    InstantQuery.gpuFbFree: gpuFbFreeLast,
    InstantQuery.gpuPower: gpuPowerLast,
    InstantQuery.gpuSmClock: gpuSmClockLast,
    InstantQuery.gpuMemClock: gpuMemClockLast,
    InstantQuery.gpuMemTemp: gpuMemTempLast,
    InstantQuery.gpuAgeDeep: gpuAgeDeep,
  };

  static String cpuFor(String instance) {
    final i = _safe(instance);
    return '100 - (avg by (instance) '
        '(rate(node_cpu_seconds_total{instance="$i",mode="idle"}[2m])) * 100)';
  }

  /// Memory in use, in bytes. The NODES chart plots this rather than the
  /// percentage: on a guest you want to see how many gigabytes a workload
  /// actually took, and a percentage hides the size of the box it ran on.
  static String memUsedBytesFor(String instance) {
    final i = _safe(instance);
    return 'node_memory_MemTotal_bytes{instance="$i"} - '
        'node_memory_MemAvailable_bytes{instance="$i"}';
  }

  static String memPctFor(String instance) {
    final i = _safe(instance);
    return '(1 - (node_memory_MemAvailable_bytes{instance="$i"} / '
        'node_memory_MemTotal_bytes{instance="$i"})) * 100';
  }

  /// Labelled hwmon history for one target, for the NODES temperature chart.
  static String hwmonTempFor(String instance) {
    final i = _safe(instance);
    return 'node_hwmon_temp_celsius{instance="$i"} '
        '* on(instance,chip,sensor) group_left(label) node_hwmon_sensor_label';
  }

  /// Raw DCGM history for one target. No freshness filter and no
  /// `last_over_time`: a range query already carries its own timestamps, and a
  /// gap in the matrix is exactly what the chart should show.
  static String gpuTempFor(String instance) =>
      'DCGM_FI_DEV_GPU_TEMP{instance="${_safe(instance)}"}';

  static String gpuUtilFor(String instance) =>
      'DCGM_FI_DEV_GPU_UTIL{instance="${_safe(instance)}"}';

  static String _safe(String s) => s.replaceAll(RegExp(r'["\\\n]'), '');
}

File: src/prom/queries.rs
//! Every PromQL expression the app issues, in one place.
use crate::config::{Config, GPU_STALE_AFTER};
#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
pub enum InstantQuery {
    Up,
    CpuBusy,
    Cores,
    MemTotal,
    MemAvailable,
    SwapTotal,
    SwapFree,
    Load1,
    Load5,
    Load15,
    BootTime,
    FsSize,
    FsAvail,
    NetRx,
    NetTx,
    AllTemps,
    CpuIoWait,
    GpuTemp,
    GpuUtil,
    GpuFbUsed,
    GpuFbFree,
    GpuPower,
    GpuSmClock,
    GpuMemClock,
    GpuMemTemp,
    GpuAgeDeep,
    GpuAgeFresh,
}
pub const CPU_BUSY: &str =
    "100 - (avg by (instance) (rate(node_cpu_seconds_total{mode=\"idle\"}[2m])) * 100)";
pub const CPU_IO_WAIT: &str =
    "avg by (instance) (rate(node_cpu_seconds_total{mode=\"iowait\"}[2m])) * 100";
pub const CORES: &str = "count by (instance) (node_cpu_seconds_total{mode=\"idle\"})";
pub const MEM_TOTAL: &str = "node_memory_MemTotal_bytes";
pub const MEM_AVAILABLE: &str = "node_memory_MemAvailable_bytes";
pub const SWAP_TOTAL: &str = "node_memory_SwapTotal_bytes";
pub const SWAP_FREE: &str = "node_memory_SwapFree_bytes";
pub const LOAD1: &str = "node_load1";
pub const LOAD5: &str = "node_load5";
pub const LOAD15: &str = "node_load15";
pub const BOOT_TIME: &str = "node_boot_time_seconds";
pub const FS_SIZE: &str = "node_filesystem_size_bytes{mountpoint=\"/\"}";
pub const FS_AVAIL: &str = "node_filesystem_avail_bytes{mountpoint=\"/\"}";
pub const UP: &str = "up";
/// Labelled hwmon temperatures give readable names such as Tctl and Tccd1.
pub const CPU_TEMP: &str =
    "node_hwmon_temp_celsius * on(instance,chip,sensor) group_left(label) node_hwmon_sensor_label";
/// A left join: labelled series win while unmatched raw channels stay visible.
pub const ALL_TEMPS: &str = "(node_hwmon_temp_celsius * on(instance,chip,sensor) group_left(label) node_hwmon_sensor_label) or on(instance,chip,sensor) node_hwmon_temp_celsius";
/// Age while temperature remains in the normal Prometheus lookback window.
pub const GPU_AGE_FRESH: &str = "time() - timestamp(DCGM_FI_DEV_GPU_TEMP)";
/// An exporter can have been down for hours or days; search seven days back.
pub const GPU_AGE_DEEP: &str = "time() - max_over_time(timestamp(DCGM_FI_DEV_GPU_TEMP)[7d:5m])";
pub const RANGE_CPU: &str =
    "100 - (avg by (instance) (rate(node_cpu_seconds_total{mode=\"idle\"}[2m])) * 100)";
pub const RANGE_MEM_PCT: &str =
    "(1 - (node_memory_MemAvailable_bytes / node_memory_MemTotal_bytes)) * 100";
pub const RANGE_TEMP_CPU: &str =
    "node_hwmon_temp_celsius * on(instance,chip,sensor) group_left(label) node_hwmon_sensor_label";
pub const RANGE_TEMP_GPU: &str = "DCGM_FI_DEV_GPU_TEMP";
pub const RANGE_GPU_UTIL: &str = "DCGM_FI_DEV_GPU_UTIL";
/// End-to-end throughput per path, not NIC traffic. The exporter runs tens of
/// minutes apart, so hold each measurement until the next one arrives.
pub const RANGE_SPEEDTEST_DOWN: &str = "last_over_time(speedtest_download_bits_per_second[1h])";
/// Both directions are measured in one run and honestly share one axis.
pub const RANGE_SPEEDTEST_UP: &str = "last_over_time(speedtest_upload_bits_per_second[1h])";
pub const RANGE_SPEEDTEST_LATENCY: &str = "last_over_time(speedtest_latency_seconds[1h])";

/// Fresh selectors reject lookback-window ghosts; history is queried separately.
pub fn fresh_gpu(metric: &str) -> String {
    format!(
        "{metric} and on(instance,gpu,UUID) (time() - timestamp({metric}) < {})",
        GPU_STALE_AFTER.as_secs()
    )
}
pub fn last_gpu(metric: &str) -> String {
    format!("last_over_time({metric}[7d])")
}
/// Escape the configurable regex as a PromQL string without changing its meaning.
fn quote_regex(s: &str) -> String {
    serde_json::to_string(s).expect("string serialization")
}
pub fn net_rx(cfg: &Config) -> String {
    format!(
        "sum by (instance) (rate(node_network_receive_bytes_total{{device!~{}}}[2m]))",
        quote_regex(&cfg.net_device_exclude)
    )
}
pub fn net_tx(cfg: &Config) -> String {
    format!(
        "sum by (instance) (rate(node_network_transmit_bytes_total{{device!~{}}}[2m]))",
        quote_regex(&cfg.net_device_exclude)
    )
}
pub fn range_net_rx(cfg: &Config) -> String {
    net_rx(cfg)
}
pub fn gpu_temp() -> String {
    fresh_gpu("DCGM_FI_DEV_GPU_TEMP")
}
pub fn gpu_temp_last() -> String {
    last_gpu("DCGM_FI_DEV_GPU_TEMP")
}
pub fn gpu_mem_temp() -> String {
    fresh_gpu("DCGM_FI_DEV_MEMORY_TEMP")
}
pub fn gpu_mem_temp_last() -> String {
    last_gpu("DCGM_FI_DEV_MEMORY_TEMP")
}
pub fn gpu_util() -> String {
    fresh_gpu("DCGM_FI_DEV_GPU_UTIL")
}
pub fn gpu_util_last() -> String {
    last_gpu("DCGM_FI_DEV_GPU_UTIL")
}
pub fn gpu_fb_used() -> String {
    fresh_gpu("DCGM_FI_DEV_FB_USED")
}
pub fn gpu_fb_used_last() -> String {
    last_gpu("DCGM_FI_DEV_FB_USED")
}
pub fn gpu_fb_free() -> String {
    fresh_gpu("DCGM_FI_DEV_FB_FREE")
}
pub fn gpu_fb_free_last() -> String {
    last_gpu("DCGM_FI_DEV_FB_FREE")
}
pub fn gpu_power() -> String {
    fresh_gpu("DCGM_FI_DEV_POWER_USAGE")
}
pub fn gpu_power_last() -> String {
    last_gpu("DCGM_FI_DEV_POWER_USAGE")
}
pub fn gpu_sm_clock() -> String {
    fresh_gpu("DCGM_FI_DEV_SM_CLOCK")
}
pub fn gpu_sm_clock_last() -> String {
    last_gpu("DCGM_FI_DEV_SM_CLOCK")
}
pub fn gpu_mem_clock() -> String {
    fresh_gpu("DCGM_FI_DEV_MEM_CLOCK")
}
pub fn gpu_mem_clock_last() -> String {
    last_gpu("DCGM_FI_DEV_MEM_CLOCK")
}
/// Production and live smoke tests share this list so neither drifts behind.
pub fn instant_poll_queries(cfg: &Config) -> Vec<(InstantQuery, String)> {
    vec![
        (InstantQuery::Up, UP.into()),
        (InstantQuery::CpuBusy, CPU_BUSY.into()),
        (InstantQuery::Cores, CORES.into()),
        (InstantQuery::MemTotal, MEM_TOTAL.into()),
        (InstantQuery::MemAvailable, MEM_AVAILABLE.into()),
        (InstantQuery::SwapTotal, SWAP_TOTAL.into()),
        (InstantQuery::SwapFree, SWAP_FREE.into()),
        (InstantQuery::Load1, LOAD1.into()),
        (InstantQuery::Load5, LOAD5.into()),
        (InstantQuery::Load15, LOAD15.into()),
        (InstantQuery::BootTime, BOOT_TIME.into()),
        (InstantQuery::FsSize, FS_SIZE.into()),
        (InstantQuery::FsAvail, FS_AVAIL.into()),
        (InstantQuery::NetRx, net_rx(cfg)),
        (InstantQuery::NetTx, net_tx(cfg)),
        (InstantQuery::AllTemps, ALL_TEMPS.into()),
        (InstantQuery::CpuIoWait, CPU_IO_WAIT.into()),
        (InstantQuery::GpuTemp, gpu_temp()),
        (InstantQuery::GpuUtil, gpu_util()),
        (InstantQuery::GpuFbUsed, gpu_fb_used()),
        (InstantQuery::GpuFbFree, gpu_fb_free()),
        (InstantQuery::GpuPower, gpu_power()),
        (InstantQuery::GpuSmClock, gpu_sm_clock()),
        (InstantQuery::GpuMemClock, gpu_mem_clock()),
        (InstantQuery::GpuMemTemp, gpu_mem_temp()),
        (InstantQuery::GpuAgeFresh, GPU_AGE_FRESH.into()),
    ]
}
/// Expensive diagnostics, only for dcgm targets whose up is currently zero.
pub fn gpu_fallback_queries(cfg: &Config) -> Vec<(InstantQuery, String)> {
    let _ = cfg;
    vec![
        (InstantQuery::GpuTemp, gpu_temp_last()),
        (InstantQuery::GpuUtil, gpu_util_last()),
        (InstantQuery::GpuFbUsed, gpu_fb_used_last()),
        (InstantQuery::GpuFbFree, gpu_fb_free_last()),
        (InstantQuery::GpuPower, gpu_power_last()),
        (InstantQuery::GpuSmClock, gpu_sm_clock_last()),
        (InstantQuery::GpuMemClock, gpu_mem_clock_last()),
        (InstantQuery::GpuMemTemp, gpu_mem_temp_last()),
        (InstantQuery::GpuAgeDeep, GPU_AGE_DEEP.into()),
    ]
}

/// Strip quotes, backslashes and newlines from an instance label, as in Dart.
pub fn safe(s: &str) -> String {
    s.chars()
        .filter(|c| !matches!(c, '"' | '\\' | '\n'))
        .collect()
}
pub fn cpu_for(instance: &str) -> String {
    let i = safe(instance);
    format!(
        "100 - (avg by (instance) (rate(node_cpu_seconds_total{{instance=\"{i}\",mode=\"idle\"}}[2m])) * 100)"
    )
}
/// Bytes tell how much memory the workload took; percent hides the box's size.
pub fn mem_used_bytes_for(instance: &str) -> String {
    let i = safe(instance);
    format!(
        "node_memory_MemTotal_bytes{{instance=\"{i}\"}} - node_memory_MemAvailable_bytes{{instance=\"{i}\"}}"
    )
}
pub fn mem_pct_for(instance: &str) -> String {
    let i = safe(instance);
    format!(
        "(1 - (node_memory_MemAvailable_bytes{{instance=\"{i}\"}} / node_memory_MemTotal_bytes{{instance=\"{i}\"}})) * 100"
    )
}
pub fn hwmon_temp_for(instance: &str) -> String {
    let i = safe(instance);
    format!(
        "node_hwmon_temp_celsius{{instance=\"{i}\"}} * on(instance,chip,sensor) group_left(label) node_hwmon_sensor_label"
    )
}
/// Raw history keeps its own timestamps; gaps should remain gaps.
pub fn gpu_temp_for(instance: &str) -> String {
    format!("DCGM_FI_DEV_GPU_TEMP{{instance=\"{}\"}}", safe(instance))
}
pub fn gpu_util_for(instance: &str) -> String {
    format!("DCGM_FI_DEV_GPU_UTIL{{instance=\"{}\"}}", safe(instance))
}
